src/metadata.rs

```diff
@@ -361,10 +458,12 @@ impl AuthorizationMetadata {
         JE: JweContentEncryptionAlgorithm<JT>,
         JA: JweKeyManagementAlgorithm + Clone,
     {
-        let issuer_url = issuer_metadata
-            .authorization_server
-            .clone()
-            .unwrap_or(issuer_metadata.credential_issuer.clone());
+        let issuer_url = (match &issuer_metadata.authorization_servers {
+            Some(v) => v.clone().into_iter().next(),
```

Comment:
After a quick look, I can't find how this is supposed to be interpreted, but surely it can't simply be "pick the first one" otherwise what's the point of having an array?

Reply:
Yup, I forgot to add a TODO here so I missed this in my review before pushing, I have to go back to the spec and fix this and the other similar section.

Reply:
From https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0-ID1.html#name-credential-issuer-metadata-p it looks like this function could accept an optional `grant_type` parameter to filter servers based on it

Reply:
Is something like this what we should be doing then?
I'm not sure I understood this section 100%.

```rust
    pub fn discover<HC, RE, CM, JT, JE, JA>(
        issuer_metadata: &IssuerMetadata<CM, JT, JE, JA>,
        grant_type: Option<CoreGrantType>,
        http_client: HC,
    ) -> Result<Self, DiscoveryError<RE>>
    where
        HC: Fn(HttpRequest) -> Result<HttpResponse, RE>,
        RE: std::error::Error + 'static,
        CM: CredentialMetadataProfile,
        JT: JsonWebKeyType,
        JE: JweContentEncryptionAlgorithm<JT>,
        JA: JweKeyManagementAlgorithm + Clone,
    {
        let issuer_url = (match (&issuer_metadata.authorization_servers, grant_type) {
            // TODO: respond with the appropriate authorization server
            (Some(v), Some(grant_type)) if !v.is_empty() => v
                .iter()
                .map(|auth_server| {
                    let auth_server_url = auth_server
                        .join(AUTHORIZATION_METADATA_URL_SUFFIX)
                        .map_err(DiscoveryError::UrlParse)?;

                    http_client(Self::discovery_request(auth_server_url))
                        .map_err(DiscoveryError::Request)
                        .and_then(|http_response| {
                            Self::discovery_response(auth_server, http_response)
                        })
                })
                .collect::<Result<Vec<Self>, _>>()
                .and_then(|disc_responses| {
                    match disc_responses.into_iter().find(|disc_response| {
                        match disc_response.0.grant_types_supported() {
                            Some(gts) => gts.iter().any(|gt| *gt == grant_type),
                            None => false,
                        }
                    }) {
                        Some(server) => Ok(Some(server.0.issuer().to_owned())),
                        _ => Err(DiscoveryError::Other(
                            "Failed to filter authorization server".to_string(),
                        )),
                    }
                }),
            (Some(v), _) if v.is_empty() => Ok(None),
            _ => Ok(None),
        })?
        .unwrap_or(issuer_metadata.credential_issuer.clone());

        let discovery_url = issuer_url
            .join(AUTHORIZATION_METADATA_URL_SUFFIX)
            .map_err(DiscoveryError::UrlParse)?;

        http_client(Self::discovery_request(discovery_url))
            .map_err(DiscoveryError::Request)
            .and_then(|http_response| Self::discovery_response(&issuer_url, http_response))
    }
```

Reply:
Yes, this reads correct to me.

Reply:
The logic looks good but:
- it could be optimised to only iteratively fetch the metadata and break out at the first one that's good, and avoid fetching it again at the end
- log errors (with `tracing`) when the discovery of authorization servers fail instead of failing the whole discovery (this is arguable, but it does seem to match the spirit of the specifications IMO)

Reply:
Is something like this what you had in mind?
I added an optional parameter for a logging function similar to how we receive an `http_client`, as we didn't have `tracing` or logging anywhere else in the library.

```rust
    pub fn discover<LF, HC, RE, CM, JT, JE, JA>(
        credential_issuer_metadata: &CredentialIssuerMetadata<CM, JT, JE, JA>,
        grant_type: Option<CoreGrantType>,
        http_client: HC,
        logging_fn: Option<LF>,
    ) -> Result<Self, DiscoveryError<RE>>
    where
        LF: Fn(&DiscoveryError<RE>),
        HC: Fn(HttpRequest) -> Result<HttpResponse, RE>,
        RE: std::error::Error + 'static,
        CM: CredentialMetadataProfile,
        JT: JsonWebKeyType,
        JE: JweContentEncryptionAlgorithm<JT>,
        JA: JweKeyManagementAlgorithm + Clone,
    {
        let log_and_continue = |e: DiscoveryError<RE>| -> DiscoveryError<RE> {
            if let Some(ref logging_fn) = logging_fn {
                logging_fn(&e);
            }
            e
        };

        match (
            &credential_issuer_metadata.authorization_servers,
            grant_type,
        ) {
            (Some(v), Some(grant_type)) if !v.is_empty() => v
                .iter()
                .map(|auth_server| {
                    let auth_server_url = auth_server
                        .join(AUTHORIZATION_METADATA_URL_SUFFIX)
                        .map_err(DiscoveryError::UrlParse)
                        .map_err(log_and_continue)?;

                    let response = http_client(Self::discovery_request(auth_server_url))
                        .map_err(DiscoveryError::Request)
                        .map_err(log_and_continue)
                        .and_then(|http_response| {
                            Self::discovery_response(auth_server, http_response)
                        })?;

                    Ok(response.0.grant_types_supported().and_then(|gts| {
                        if gts.iter().any(|gt| *gt == grant_type) {
                            Some(response.clone())
                        } else {
                            None
                        }
                    }))
                })
                .take_while(|s| !matches!(s, Ok(Some(_))))
                .next()
                .ok_or(DiscoveryError::Other(
                    "failed to select an authorization server: no matching grant_type".to_string(),
                ))??
                .ok_or(DiscoveryError::Other(
                    "failed to select an authorization server: no matching grant_type".to_string(),
                )),

            _ => {
                let issuer_url = credential_issuer_metadata.credential_issuer.clone();

                let discovery_url = issuer_url
                    .join(AUTHORIZATION_METADATA_URL_SUFFIX)
                    .map_err(DiscoveryError::UrlParse)?;

                http_client(Self::discovery_request(discovery_url))
                    .map_err(DiscoveryError::Request)
                    .and_then(|http_response| Self::discovery_response(&issuer_url, http_response))
            }
        }
    }
```

Reply:
I'm also now questioning whether an empty `authorization_servers` array should error or default to `credential_issuer`.

Reply:
Yeah, the specs don't specify that the array must be non-empty. But obviously considering we can't do much with an empty array, I would vote to then default to `credential_issuer` (which would be the same behaviour if no `authorization_server` were to support the `grant_type` so at least it's consistent in this way)

Reply:
Won't this have a possible side effect of allowing badly/wrongly configured servers to end up relying on unspecified behavior to work?

Reply:
Also in the code I sent above, if no authorization server from the array supports the `grant_type`, the call will return an error.